### PR TITLE
splice: Reestablish when commit or sig sends fail

### PR DESCRIPTION
**Pull request by ddustin**
Adds tests for when the connection fails during
1) splice tx_signature
2) splice commitment_signed

Fleshed out the reestablish flow for these two cases and implemented the fixes to make these reestablish flows work.

Fixes https://github.com/ElementsProject/lightning/issues/6703

Changelog-Fixed: Implemented splicing restart logic for tx_signature and commitment_signed.

### CODE REVIEW

**Review by vincenzopalazzo — COMMENTED** [review]
Looks like that test are failing

```
test_splice passed 1 out of the required 1 times. Success!

===End Flaky Test Report===
=========================== short test summary info ============================
FAILED tests/test_opening.py::test_rbf_reconnect_tx_construct - TimeoutError: Unable to find "[re.compile('dualopend daemon died before signed PSBT returned')]" in logs.
FAILED tests/test_splicing_disconnect.py::test_splice_disconnect_sig - TimeoutError: Unable to find "[re.compile('CHANNELD_AWAITING_SPLICE to CHANNELD_NORMAL')]" in logs.
===== 2 failed, 105 passed, 582 skipped, 33 warnings in 699.05s (0:11:39) ======
```

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/6840/commits/50bcd8932006dfb9ba3a53016b5109a06754bea6

File: channeld/channeld.c [hunk 1]
```diff
@@ -2824,10 +2834,11 @@ static struct wally_psbt *next_splice_step(const tal_t *ctx,
...
 
-static const u8 *peer_expect_msg(const tal_t *ctx,
-				 struct peer *peer,
-				 enum peer_wire expect_type,
-				 enum peer_wire second_allowed_type)
+static const u8 *peer_expect_msg_three(const tal_t *ctx,
+				       struct peer *peer,
+				       enum peer_wire expect_type,
+				       enum peer_wire second_allowed_type,
+				       enum peer_wire third_allowed_type)
```
Comment (rustyrussell) [inline channeld/channeld.c:2841]: You never check third_allowed_type?

**Review by rustyrussell — CHANGES_REQUESTED** [review]
I'm going to accept this since we're close to release and it only effects experimental.  However, it's a giant blob of unreviewable changes, which is not nice :(

At least one thing is wrong, though.

**Comment by vincenzopalazzo** [comment]
This is introducing breaking change so @nepet drop for the milestone, it is breaking VLS with https://github.com/ElementsProject/lightning/issues/6870 so we should drop it from the current release IMHO too

**Review by ksedgwic — APPROVED** [review]
Investigated the "commitment mismatch" VLS was seeing with `tests/test_splicing_disconnect.py::test_splice_disconnect_commit`, looks like this is not a bug and we can easily handle in VLS: https://gitlab.com/lightning-signer/vls-hsmd/-/merge_requests/132#note_1654982750

This PR looks good from VLS

**Comment by nepet** [comment]
Squashed fixup commit

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/commit/b54af9f8f4e24685a368bb5185aa068fbd573d6d

File: channeld/channeld.c [hunk 2]
```diff
@@ -3323,12 +3379,69 @@ static void update_view_from_inflights(struct peer *peer)
...
+	return true;
+}
+
+/* Called to finish an ongoing splice OR on restart from chanenl_reestablish. */
 static void resume_splice_negotiation(struct peer *peer,
-				      struct inflight *inflight,
-				      bool skip_commitments,
-				      enum tx_role our_role)
+				      bool send_commitments,
+				      bool recv_commitments,
```
Comment (niftynei) [inline channeld/channeld.c:3437]: nit: maybe call this 'expect_rcv_commitments'?

File: channeld/channeld.c [hunk 3]
```diff
@@ -3358,10 +3480,20 @@ static void resume_splice_negotiation(struct peer *peer,
...
+						    last_inflight_index(peer),
+						    send_commitments,
+						    recv_commitments,
+						    &msg_received);
+
+	if (check_tx_abort(peer, msg_received))
+		return;
 
+	if (their_commit) {
+		if (inflight->last_tx != their_commit->tx)
```
Comment (niftynei) [inline channeld/channeld.c:3495]: in what cases would these not be the same?

File: channeld/channeld.c [hunk 4]
```diff
@@ -3407,155 +3545,198 @@ static void resume_splice_negotiation(struct peer *peer,
...
-	msg = towire_channeld_splice_confirmed_signed(tmpctx, final_tx,
-						      chan_output_index);
-	wire_sync_write(MASTER_FD, take(msg));
+	if (recv_signature) {
+		msg = towire_channeld_splice_confirmed_signed(tmpctx, final_tx,
+							      chan_output_index);
+		wire_sync_write(MASTER_FD, take(msg));
 
-	send_channel_update(peer, true);
+		send_channel_update(peer, true);
```
Comment (niftynei) [inline channeld/channeld.c:3738]: can we get rid of this? what's it for here?

File: channeld/channeld.c [hunk 5]
```diff
@@ -3749,7 +3932,7 @@ static void splice_accepter(struct peer *peer, const u8 *inmsg)
 
 	peer->splice_state->count++;
 
-	resume_splice_negotiation(peer, new_inflight, false, TX_ACCEPTER);
+	resume_splice_negotiation(peer, true, true, true, true);
```
Comment (niftynei) [inline channeld/channeld.c:3935]: nit: could you compact these args into a bitfield/enum?

File: channeld/channeld.c [hunk 6]
```diff
@@ -2824,17 +2841,19 @@ static struct wally_psbt *next_splice_step(const tal_t *ctx,
 	return ictx->desired_psbt;
 }
 
-static const u8 *peer_expect_msg(const tal_t *ctx,
-				 struct peer *peer,
-				 enum peer_wire expect_type,
-				 enum peer_wire second_allowed_type)
+static const u8 *peer_expect_msg_three(const tal_t *ctx,
```
Comment (niftynei) [inline channeld/channeld.c:2844]: nit: switch statement?

File: tests/test_splicing_disconnect.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    l1 = node_factory.get_node(disconnect=disconnect,
+                               options={'experimental-splicing': None, 'dev-no-reconnect': None},
+                               may_reconnect=True)
+    l2 = node_factory.get_node(options={'experimental-splicing': None}, may_reconnect=True)
+    l1.openchannel(l2, 1000000)
+
+    chan_id = l1.get_channel_id(l2)
+
+    # add extra sats to pay fee
+    funds_result = l1.rpc.fundpsbt("109000sat", "slow", 166, excess_as_change=True)
```
Comment (niftynei) [inline tests/test_splicing_disconnect.py:29]: wen splice cmd?? :eyes:

File: tests/test_splicing_disconnect.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    print("Killing l1 without sending WIRE_TX_SIGNATURES")
+    l1.daemon.kill()
+
+    # Restart l1, without disconnect stuff.
+    del l1.daemon.opts['dev-no-reconnect']
+    del l1.daemon.opts['dev-disconnect']
+
+    # Should reconnect, and reestablish the splice.
+    l1.start()
```
Comment (niftynei) [inline tests/test_splicing_disconnect.py:46]: you could also just call `connect` again?

File: tests/test_splicing_disconnect.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    bitcoind.generate_block(6, wait_for_mempool=1)
+
+    l1.daemon.wait_for_log(r'CHANNELD_AWAITING_SPLICE to CHANNELD_NORMAL')
+    l2.daemon.wait_for_log(r'CHANNELD_AWAITING_SPLICE to CHANNELD_NORMAL')
+    l1.daemon.wait_for_log(r'private channel announcement from channeld for ' + first_scid(l1, l2))
+
+    inv = l2.rpc.invoice(10**2, '3', 'no_3')
+    l1.rpc.pay(inv['bolt11'])
+
+    # Check that the splice doesn't generate a unilateral close transaction
```
Comment (niftynei) [inline tests/test_splicing_disconnect.py:61]: is there another signal we can check for?

File: tests/test_splicing_disconnect.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                               may_reconnect=True)
+    l1.openchannel(l2, 1000000)
+
+    chan_id = l1.get_channel_id(l2)
+
+    # add extra sats to pay fee
+    funds_result = l1.rpc.fundpsbt("109000sat", "slow", 166, excess_as_change=True)
+
+    result = l1.rpc.splice_init(chan_id, 100000, funds_result['psbt'])
+    print("l1 splice_update")
```
Comment (niftynei) [inline tests/test_splicing_disconnect.py:82]: can we get rid of the prints? :grin:

File: tests/test_splicing_disconnect.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    result = l1.rpc.signpsbt(result['psbt'])
+    print("l1 splice_signed")
+
+    executor.submit(l1.rpc.splice_signed, chan_id, result['signed_psbt'])
+
+    print("l2 waiting for dev_disconnect msg")
+
+    l2.daemon.wait_for_log(r'dev_disconnect: \+WIRE_COMMITMENT_SIGNED')
+
+    print("Killing l2 without sending WIRE_COMMITMENT_SIGNED")
```
Comment (niftynei) [inline tests/test_splicing_disconnect.py:94]: nit: assert we're in the state expected before killing + reconnecting

File: tests/test_splicing_disconnect.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    l1.daemon.wait_for_log(r'CHANNELD_AWAITING_SPLICE to CHANNELD_NORMAL')
+    l2.daemon.wait_for_log(r'CHANNELD_AWAITING_SPLICE to CHANNELD_NORMAL')
+    l1.daemon.wait_for_log(r'private channel announcement from channeld for ' + first_scid(l1, l2))
+
+    inv = l2.rpc.invoice(10**2, '3', 'no_3')
+    l1.rpc.pay(inv['bolt11'])
+
+    # Check that the splice doesn't generate a unilateral close transaction
+    time.sleep(5)
+    assert l1.db_query("SELECT count(*) as c FROM channeltxs;")[0]['c'] == 0
```
Comment (niftynei) [inline tests/test_splicing_disconnect.py:119]: one other test i would add tests the tx-abort pathway?

**Review by niftynei — APPROVED** [review]
Bunch of nits, but nothing major. Nice job getting this turned around so quickly.

More tests would be nice??

**Pull request merged**